compiler/rustc_mir_transform/src/coverage/spans.rs

```diff
@@ -321,30 +321,17 @@ impl<'a> CoverageSpansGenerator<'a> {
             }
         }
 
-        let prev = self.take_prev();
-        debug!("    AT END, adding last prev={prev:?}");
-
         // Take `pending_dups` so that we can drain it while calling self methods.
         // It is never used as a field after this point.
         for dup in std::mem::take(&mut self.pending_dups) {
             debug!("    ...adding at least one pending dup={:?}", dup);
             self.push_refined_span(dup);
         }
 
-        // Async functions wrap a closure that implements the body to be executed. The enclosing
-        // function is called and returns an `impl Future` without initially executing any of the
-        // body. To avoid showing the return from the enclosing function as a "covered" return from
-        // the closure, the enclosing function's `TerminatorKind::Return`s `CoverageSpan` is
-        // excluded. The closure's `Return` is the only one that will be counted. This provides
-        // adequate coverage, and more intuitive counts. (Avoids double-counting the closing brace
-        // of the function body.)
-        let body_ends_with_closure = if let Some(last_covspan) = self.refined_spans.last() {
-            last_covspan.is_closure && last_covspan.span.hi() == self.body_span.hi()
-        } else {
-            false
-        };
-
-        if !body_ends_with_closure {
+        // There is usually a final span remaining in `prev` after the loop ends,
+        // so add it to the output as well.
+        if let Some(prev) = self.some_prev.take() {
```

Comment:
Switching from `take_prev` (`unwrap`) to `if let` here might also have the happy side-effect of fixing #118643.

Reply:
Could you add a test for this ?

Reply:
I can try to write one, but I don't have a minimal repro to work from, and triggering the problem is likely to require some tricky interactions with macro expansion. So it might be quite difficult to come up with a suitable test.

Reply:
I was able to come up with a test, so I've added it.